### PR TITLE
Add merged pull request reminder 📎

### PR DESCRIPTION
**Pull request by alisdair**
Untested (untestable?) GitHub actions workflow adding a comment to try to remind us to update the changelog when merging a PR.

### CODE REVIEW

File: .github/workflows/merged-pr.yml
Comment (crw) [inline .github/workflows/merged-pr.yml:22]: Can we link to instructions as to how to write an appropriate changelog message? Or is that just kinda obvious?

Comment (crw) [reply .github/workflows/merged-pr.yml:22]: Don't let this block the PR. Maybe for a future change once we see how this works.

**Comment by apparentlymart** [comment]
When we were originally discussing this I had in my mind a notion that all of our "PR workflow" stuff ought to be together in a single system so that we could avoid having many separate GitHub actions all trying to trample on the same turf, and thus I was proposing to add this to my rather-larger-and-still-unfinished "label-and-comment-and-backport-bonanza" workflow.

However, now I see this simple approach in isolation it feels more comfortable to me for this one to stay separate, since what it is doing is straightforward and isolated from all other work. (Specifically, it doesn't interact with any _shared mutable state_ of a pull request: no label reading/writing, and merging is an irreversible action.

With that said then, this looks good to me but I have a few small notes based on my recent experiences writing a similar PR-reactive workflow:
- This seems like a good candidate to be run on `pull_request_target` rather than `pull_request`, which means that GitHub will use the workflow as defined in the _target_ branch rather than the workflow as defined in the _source_ branch. In return for that, the workflow can have automatic access to a privileged token without any special secrets configuration, because changes to the workflow will be subject to PR _before_ they take effect.
- Building on that, I think the limited scope of this workflow means it would work fine with a restricted-permissions access token declared with something like the following at the top level:

    ```yaml
    permissions:
      pull-requests: write
    ```

    (Caveat: I've not actually tested whether that grants sufficient access for _commenting_, but I don't see anything else in [the relevant docs](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs#overview) which looks like it might be relevant.)

**Comment by apparentlymart** [comment]
Oh, also: I set up a private repository to test my own workflow work in, which you're welcome to borrow for testing this if you want. Let me know via an internal communication channel if you'd like to hear more.

File: .github/workflows/merged-pr.yml
Comment (apparentlymart) [reply .github/workflows/merged-pr.yml:22]: We're intending this message to be for the person who did the merge (someone who works at HashiCorp) rather than the person who submitted the PR, and so I think we're assuming they already know what they need to do and we're basically just creating a GitHub notification to remind them to do it.

Given that though, it might be worth making it clearer in the message text that its audience _is_ the maintainer who merged it, so that if we're tardy in doing the changelog update we don't make the original submitter feel like they must e.g. open _another_ PR to suggest a changelog addition, or whatever.

Comment (alisdair) [reply .github/workflows/merged-pr.yml]: > Can we link to instructions as to how to write an appropriate changelog message?

Good thought! I don't think we currently _have_ any such instructions and I don't currently have the capacity to try writing them. When they do exist, we should update this hook to point to them.

**Comment by alisdair** [comment]
Thanks for the suggestions! I've updated the workflow and successfully tested on a private repo of my own.

File: .github/workflows/merged-pr.yml
Comment (crw) [reply .github/workflows/merged-pr.yml]: Thanks! I was thinking the "how to write a changelog entry" would be targeted towards an internal contributor, although given my role I suppose I should be thinking more of our external audience. 😅

**Review by apparentlymart — APPROVED** [review]

**Pull request merged**
